[MRG+2] Add common test for set_params behavior

absolutelyNoWarranty opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
Fix #7738  
#### What does this implement/fix? Explain your changes.

~~Test `set_params` by adding `estimator.set_params(**params)` to `check_parameters_default_constructible`~~
Test `set_params` by adding a `check_set_params` to the `estimator_checks` module.
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

absolutelyNoWarranty committed:
Add set_param check and test

Add check_set_params to test_common
Add assert_is

amueller commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -66,6 +75,9 @@ def test_check_estimator():
     # check that we have a set_params and can clone
     msg = "it does not implement a 'get_params' methods"
     assert_raises_regex(TypeError, msg, check_estimator, object)
+    # check that properties can be set
```
wow I kinda forgot that I wrote this test. This looks great!

amueller commented:
Maybe add a stronger test that ensures idempotence as you pointed out? You could have a getter and setter but they are not inverse of each other, like a ``+1`` in one or both of them.

absolutelyNoWarranty commented:
Is it against sklearn's conventions if a set_param fails due to TypeError, etc? If the getter and setter do processing, then they can fail for all kinds of reasons.

amueller commented:
There is no conventions on what errors can be raised where, I think. I'm not quite sure what you're getting at.

absolutelyNoWarranty commented:
If `p` is a property setter with logic (i.e. an addition operation), it could raise Exceptions. Wouldn't that be accidentally doing input validation at `init` time? In this case at `init` non numeric values would raise a TypeError.

Currently the contributor guidelines say:
> The reason for postponing the validation is that the same validation would have to be performed in set_params, which is used in algorithms like GridSearchCV.

What I am wondering is if property-params are subject to this rule, or not.

Finally, the reason I ask this is because I'm not sure how to test invertibliilty/idempotence in a general way except to `set_param(get_param(**params))` with a range of different kinds of values. 

Something like
```
for value in [None, -1, 1, "abc"]:
    est = est.set_params(**{'alpha':value})
    alpha_get_value = est.alpha
    assert alpha_get_value == est.set_params(**{'alpha':alpha_get_value}).get_params()['alpha']
```

If property-setters raise Exceptions then that wouldn't work.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1457,6 +1457,7 @@ def param_filter(p):
             # true for mixins
             return
         params = estimator.get_params()
+        estimator.set_params(**params)
```
Please put a comment in here to describe what this is testing.

Also: Should we do further checking that `get_params()` returns the same thing before and after this statement.

amueller replied:
yes

amueller commented:
@absolutelyNoWarranty Sorry I don't get your point about raising errors.

> What I am wondering is if property-params are subject to this rule, or not.

Again, I'm not sure what you're asking. The rule says that you shouldn't do validation in ``__init__`` because then ``set_params`` wouldn't do the validation and you'd get wrong results in GridSearchCV.
Doing validation in ``set_params`` is permissible (though we usually avoid it).
The more general rule is that calling ``__init__`` and ``set_params`` should result in the same estimator. An easy way to achieve that is to do any validation and checking and computation in ``fit``.

amueller commented:
Can you please fix the pep8 error?
Also, it would be great if you can check that ``get_params`` returns the same before and after the call to ``set_params``.

jnothman commented:
The linter still says no:
```

./sklearn/utils/estimator_checks.py:1648:1: W293 blank line contains whitespace
^
./sklearn/utils/estimator_checks.py:1649:80: E501 line too long (108 > 79 characters)
    test_values = [-np.inf, np.inf, None, -100, 100, -0.5, 0.5, 0, "", "value", ('a', 'b'), {'key':'value'}]
                                                                               ^
./sklearn/utils/estimator_checks.py:1649:99: E231 missing whitespace after ':'
    test_values = [-np.inf, np.inf, None, -100, 100, -0.5, 0.5, 0, "", "value", ('a', 'b'), {'key':'value'}]
                                                                                                  ^
./sklearn/utils/estimator_checks.py:1663:71: E502 the backslash is redundant between brackets
                             "get_params does not match set_params: " \
                                                                      ^
./sklearn/utils/estimator_checks.py:1664:80: E501 line too long (155 > 79 characters)
                             "called set_params of {0} with {1}={2} but get_params returns {3}={4}".format(name, param_name, value, param_name, get_value))
```

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1630,6 +1631,51 @@ def transform(self, X):
+                pass
+            else:
+                errmsg = ("get_params does not match set_params: "
+                          "called set_params of {0} with {1}={2} "
```
if you're gonna use explicit numbers you could have not passed param_name twice ;)

amueller commented:
@jnothman are we now doing "approve changes" instead of "[MRG + 1]"?

jnothman commented:
No, I currently haven't found an easy way to say "I don't want a cross next to my review so far because the changes I requested have been made, but I haven't done a full review of the code either". Perhaps I should just dismiss reviews more often.

amueller commented:
@jnothman Can't you change it to "comment"? (I don't actually know). I only use "comment" so far, though that is clearly also not ideal

absolutelyNoWarranty commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1630,6 +1631,51 @@ def transform(self, X):
+                estimator.set_params(**{param_name: value})
+                get_value = estimator.get_params()[param_name]
+            except:
+                # triggered some parameter validation
```
Sorry, I didn't realize there had been comments.

Here we don't error because we just want to check that `get_value` is `value` if `value` is successfully set.

absolutelyNoWarranty commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1630,6 +1631,51 @@ def transform(self, X):
+                                       param_name, get_value)
+                assert_equal(value, get_value, errmsg)
+
+        with ignore_warnings(category=DeprecationWarning):
```
Just to re-make `estimator` as in lines 1637-1641.

agramfort commented:
@absolutelyNoWarranty you need to rebase.

absolutelyNoWarranty commented:
Updated and rebased.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_set_params(name, estimator_orig):
+    if name in META_ESTIMATORS:
+        return
```
Why does this never get executed?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+            try:
+                estimator.set_params(**{param_name: value})
+                get_value = estimator.get_params()[param_name]
+            except:
```
Please use `except Exception`. bare `except` catches KeyboardInterrupt, for instance.

jnothman replied:
Or perhaps `except (ValueError, TypeError)`?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+                          "but get_params returns {1}={3}")
+                errmsg = errmsg.format(name, param_name, value,
+                                       get_value)
+                assert_equal(value, get_value, errmsg)
```
should we be asserting equality (`==`) or identity (`is`)?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+                                       get_value)
+                assert_equal(value, get_value, errmsg)
+
+        estimator = clone(estimator_orig)
```
why not put this at the beginning of the loop?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+    # values
+
+    estimator = clone(estimator_orig)
+    test_values = [-np.inf, np.inf, None,
```
I don't think we need to test so many values, if we're only checking whether they are set without modification, or an error is otherwise raised. Are we trying to trigger the estimator to validate before storing??

absolutelyNoWarranty replied:
Yes, the idea is that if there's some sort of validation it should trigger and raise an error instead of silently trying to make a correction, e.g. clipping off negative values to 0, rounding up values, etc. Obviously if this showed up in a PR somebody could just catch it by eyeballing it but I think the whole point of the original issues was not having to eyeball it. Since there's no way in advance to know what the type of the parameter is supposed to be, I couldn't think of a better way to test this other than just trying many different kinds of extreme values, for different types.

jnothman replied:
Shrug. I get what you're trying to do, but I'm not sure that's of enormous value, given that it lengthens the runtime of the test, etc (although maybe my impression that the overall test will take non-negligible time when run on all estimators should be tested).

The main thing we need to be testing is that `set_params` updates the `get_params` value. Testing that there is no validation is a bit of a side-point that we can only fully do by checking identity instead of equality.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1720,6 +1721,47 @@ def transform(self, X):
+
+            try:
+                estimator.set_params(**{param_name: value})
+                get_value = estimator.get_params()[param_name]
```
This should not be in the `try` block.

do we also want to check that all other params are unaffected?

absolutelyNoWarranty commented:
Rebased with the following changes:

Addressed review comments:
* Explicitly catch Exception, only include one statement in try
* Improve for loop
* Get rid of META_ESTIMATORS check (redundant)
* Test for identity instead of equality
* Check that other params aren't changed as well.

Replaced tests in sklearn/utils/tests/test_estimator_checks.py because of other estimator checks which have been added since the time this PR was opened. (Removed BadGetter, NoSetter; added ModifiesAnotherValue, ModifiesValueInsteadOfRaisingError, RaisesError)


Regarding runtime increase. On my laptop `check_set_param` takes roughly 15~20 seconds to run for all estimators.
```
from sklearn.utils.estimator_checks import check_set_params
from sklearn.utils.testing import all_estimators
import time

start = time.time()
for name, est_class in all_estimators():
    obj = est_class()
    check_set_params(name, obj)
end = time.time()
print(end-start)
```

jnothman commented:
It's much easier for us to track what's going on if you avoid rebasing, and instead always adding commits (including merging in master when appropriate).

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,46 @@ def transform(self, X):
+    # before and after set_params() with some fuzz
+    # values
+
+    test_values = [-np.inf, np.inf, None,
```
I'd still consider making this list shorter, because it's about such a marginal thing that we can't catch perfectly anyyway. But calling `clone` and `get_params` fewer times below will improve runtime.

absolutelyNoWarranty committed:
Drop extraneous clone

absolutelyNoWarranty committed:
Put get_params outside loop

absolutelyNoWarranty committed:
Use more readable name new_params

absolutelyNoWarranty committed:
Reset params dict every loop

absolutelyNoWarranty committed:
Make test_values much shorter

absolutelyNoWarranty committed:
Fix flake8 line too long

absolutelyNoWarranty commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,46 @@ def transform(self, X):
+           "but get_params returns {2}")
+    for param_name in params.keys():
+        estimator = clone(estimator_orig)
+        params = estimator.get_params()
```
These are here so that the params dict is reset to the defaults instead of steadily being overwritten with fuzz values.

jnothman replied:
Then set it back with the original value after fuzz?

absolutelyNoWarranty commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,46 @@ def transform(self, X):
+                assert_equal(set(params.keys()), set(_params.keys()), errmsg)
+                for k, v in _params.items():
+                    assert_is(params[k], v, errmsg)
+
```
Why is this necessary?

absolutelyNoWarranty commented:
Made requested changes.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,46 @@ def transform(self, X):
+    # before and after set_params() with some fuzz
+    # values
+
+    test_values = [-np.inf, np.inf, None]
```
I frankly don't know what to put in here. Thanks for humouring me, though! :)

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,46 @@ def transform(self, X):
+    msg = ("get_params result does not match what was passed to set_params: "
+           "called set_params of {0} with {1} "
+           "but get_params returns {2}")
+    params = estimator.get_params()
```
should we be checking that the params are now the same as before `set_params`?

jnothman reviewed (commented):
Have we reduced runtime much?

absolutelyNoWarranty committed:
Add extra dict check

absolutelyNoWarranty commented on sklearn/utils/_unittest_backport.py:
```diff
@@ -149,7 +149,7 @@ def __exit__(self, exc_type, exc_value, tb):
 
 class TestCase(unittest.TestCase):
-    longMessage = False
+    longMessage = True
```
This was set to False in #9536, but [`longMessage` is True for all `unittest.TestCase` by default.](https://docs.python.org/3/library/unittest.html#unittest.TestCase.longMessage) so I believe that was erroneous.

amueller replied:
I don't understand your comment. If it's true by default, why would someone set it to True?

absolutelyNoWarranty commented:
Added check to see that params are same as before the first set_params.
Due to reduced # of fuzz values local runtime is down to ~5s.

jnothman commented:
The reduced time is probably because you're no longer calling clone and get_params so often also...

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,42 @@ def transform(self, X):
+                pass
+            else:
+                new_params = estimator.get_params()
+                assert_equal(params.keys(), new_params.keys(), msg)
```
Not assert_is?

absolutelyNoWarranty replied:
Testing for identity doesn't work or keys even if the dictionary really is the same dictionary.
```
>>> d = {'a':1, 'b':2}
>>> d is d
True
>>> d.keys() is d.keys()
False
```

jnothman replied:
Sorry. I misread. But actually, I think that we are not assured that this equality check will work. In py2 this value is a list, in iteration order which is not to be assumed deterministic

jnothman commented on sklearn/utils/testing.py:
```diff
@@ -106,6 +106,7 @@
 assert_greater = _dummy.assertGreater
 assert_less_equal = _dummy.assertLessEqual
 assert_greater_equal = _dummy.assertGreaterEqual
+assert_is = _dummy.assertIs
```
Since we no longer use nose,
```
assert X is Y, msg
```
is actually sufficient without a helper.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,42 @@ def transform(self, X):
+    # before and after set_params() with some fuzz
+    estimator = clone(estimator_orig)
+
+    params = estimator.get_params()
```
I think we should be using deep=False, as we will test each estimator separately ...

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1915,6 +1917,42 @@ def transform(self, X):
+            try:
+                estimator.set_params(**params)
+            except Exception:
+                # triggered some parameter validation
```
In the ideal world, we might not permit this validation outside of fit either. I think we should consider raising a warning of some kind where this case is over-used.

absolutelyNoWarranty replied:
I see your point but when is the developer expected to see the warning? As part of the (massive) output of the test suite?

amueller replied:
the output of the test-suite should not be massive. There are a lot of issues in master because we were not careful about handling warnings.

amueller replied:
+1 for raising a warning (or even error?) here. Did this error for all_estimators? where? Usually each new test in ``estimator_checks`` requires fixes to the codebase. This one here is probably the one that should have been fixed. Not having any errors on the codebase is suspicious to me.

amueller replied:
For a second I thought running this on a deep estimator would fail, but I guess we avoid this by doing ``deep=False``. It would be really nice to run ``estimator_checks`` on some deep estimators though (I tried and ran into a whole host of problems)

jnothman reviewed (commented):
Getting, there, I think... But I'd like @amueller to have another go now. This feels like it's a lot of red tape.

Or perhaps we need to run it with scikit-learn-contrib estimators to check what it breaks...

absolutelyNoWarranty committed:
Use deep=False

absolutelyNoWarranty committed:
Order invariant comparison of dictionary keys

jnothman reviewed (approved):
LGTM!

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.20.rst`. There is a section for changes to estimator checks. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

jnothman commented:
The idea is that developers run check_estimator on a smaller set of objects
than we do here, and that they may be more diligent than we are about
responding to warnings!

amueller commented:
> The idea is that developers run check_estimator on a smaller set of objects than we do here, and that they may be more diligent than we are about responding to warnings!

Always the optimist. Gonna do another round now, but maybe point out what the red tape is you're concerned about?

amueller commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -142,6 +197,13 @@ def test_check_estimator():
+    assert_raises_regex(AssertionError, msg, check_estimator,
+                        ModifiesValueInsteadOfRaisingError())
+    check_estimator(RaisesError())
+    assert_raises_regex(AssertionError, msg, check_estimator,
```
can you maybe run the test on a pipeline to make sure it works on deep estimators?

absolutelyNoWarranty commented:
@jnothman @amueller 
I just realized that what should happen after an Exception is currently not checked. If parameter validation raises an Exception, should the params change still go through?
Currently this happens on some estimators:
```
>>> from sklearn.linear_model import SGDRegressor
>>> est = SGDRegressor()
>>> est.get_params()['loss']
'squared_loss'
>>> est.set_params(loss='thisdoesnotexist')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sklearn/linear_model/stochastic_gradient.py", line 78, in set_params
    self._validate_params(set_max_iter=False)
  File "sklearn/linear_model/stochastic_gradient.py", line 108, in _validate_params
    raise ValueError("The loss %s is not supported. " % self.loss)
ValueError: The loss thisdoesnotexist is not supported. 
>>> est.get_params()['loss']
'thisdoesnotexist'
```

jnothman commented:
yes, for good or bad, some estimators currently validate upon setting. I
don't think we need to be bureaucratic/frameworkish about this - such
exceptions are pretty harmless - and should focus on the essence of the PR:
testing that set_params can be used.

I'd be tempted, however, to make it only catch ValueErrors, not all
Exceptions.

absolutelyNoWarranty committed:
Raise warnings

absolutelyNoWarranty committed:
Edit whats new

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -303,3 +303,8 @@ Changes to estimator checks
   private settings apart from parameters during estimator initialization.
   :issue:`9378` by :user:`Herilalaina Rakotoarison <herilalaina>`
+
+- Added :func:`estimator_checks.check_set_params` test which checks that
```
This link won't work as we don't generate docs for estimator checks (though perhaps we should)

jnothman commented on doc/whats_new/v0.20.rst:
```diff
@@ -303,3 +303,8 @@ Changes to estimator checks
   :issue:`9378` by :user:`Herilalaina Rakotoarison <herilalaina>`
+
+- Added :func:`estimator_checks.check_set_params` test which checks that
+  `set_params` is equivalent to passing parameters in `__init__` and
```
Double backticks please.

absolutelyNoWarranty committed:
Double backticks

absolutelyNoWarranty committed:
Remove broken link

absolutelyNoWarranty committed:
fix flake8

absolutelyNoWarranty committed:
Rewrite whats_new to not start with :

absolutelyNoWarranty committed:
Fix links to check_estimator

absolutelyNoWarranty commented:
Squash and rebase?

jnothman commented:
Don't squash and rebase. We'll do that upon merge. Otherwise it makes it
hard to track what's changed.

absolutelyNoWarranty committed:
Merge branch 'master' into test_set_param

GaelVaroquaux committed:
Merge branch 'master' into test_set_param

GaelVaroquaux committed:
TST: remove assert_is

No longer needed as we are using pytests

GaelVaroquaux reviewed (approved):
LGTM.

I pushed a change remove assert_is. Once the CI are green, I will merge.
